File: apps/scala/design-pattern/test/StrategyTest.scala
package designpattern.strategy

class StrategySuite extends munit.FunSuite:

  test("HTML フォーマッタでレポートを出力する") {
    val report = Report()
    val output = report.outputReport()

    assert(output.contains("<html>"))
    assert(output.contains("<title>月次報告</title>"))
    assert(output.contains("<p>順調</p>"))
  }

  test("プレーンテキストフォーマッタに切り替える") {
    val report = Report().withFormatter(plainTextFormatter)
    val output = report.outputReport()

    assert(output.contains("**** 月次報告 ****"))
    assert(output.contains("順調"))
    assert(!output.contains("<html>"))
  }

  test("ラムダ式でカスタムフォーマッタを渡す") {
    val csvFormatter: Formatter = (title, text) => (title +: text).mkString(",")

    val report = Report(formatter = csvFormatter)
    val output = report.outputReport()

    assertEquals(output, "月次報告,順調,最高の調子")
  }

  test("case class の copy でイミュータブルに変更する") {
    val original = Report()
    val modified = original.withFormatter(plainTextFormatter)

    // 元のレポートは変更されない
    assert(original.outputReport().contains("<html>"))
    assert(!modified.outputReport().contains("<html>"))
  }

File: apps/csharp/design-pattern/tests/DesignPattern.Tests/Patterns/StrategyTest.cs
using DesignPattern.Patterns.Strategy;

namespace DesignPattern.Tests.Patterns;

public class StrategyTest
{
    private readonly string[] _text = ["Hello", "World"];

    [Fact]
    public void HtmlFormatter_ProducesHtmlOutput()
    {
        var report = new Report("Test", _text, HtmlFormatter.Format);
        var output = report.OutputReport();

        Assert.StartsWith("<html>", output);
        Assert.Contains("<p>Hello</p>", output);
    }

    [Fact]
    public void PlainTextFormatter_ProducesPlainOutput()
    {
        var report = new Report("Test", _text, PlainTextFormatter.Format);
        var output = report.OutputReport();

        Assert.Contains("***** Test *****", output);
        Assert.DoesNotContain("<html>", output);
    }

    [Fact]
    public void CanSwitchFormatterAtRuntime()
    {
        var report = new Report("Test", _text, HtmlFormatter.Format);
        Assert.Contains("<html>", report.OutputReport());

        report.Formatter = PlainTextFormatter.Format;
        Assert.Contains("*****", report.OutputReport());
    }

    [Fact]
    public void CanUseLambdaAsFormatter()
    {
        var report = new Report("Test", _text, r =>
            $"Custom: {r.Title} ({r.Text.Length} lines)");
        var output = report.OutputReport();

        Assert.Equal("Custom: Test (2 lines)", output);
    }

    [Fact]
    public void HtmlFormatter_IncludesAllLines()
    {
        var report = new Report("Test", _text, HtmlFormatter.Format);
        var output = report.OutputReport();

        Assert.Contains("<p>Hello</p>", output);
        Assert.Contains("<p>World</p>", output);
    }
}
